### PR TITLE
Use per-arch vmlinux.h v2

### PR DESCRIPTION
**Pull request by minosfuture**
Rework per-arch vmlinux.h solution
* have per-arch directory under sched/include/arch/, in which we maintain vmlinux.h symlink and real file vmlinux-{kernel_ver}-g{sha1}.h. The original sched/include/vmlinux/ folder is removed.
* update meson build `-I` option to find the new vmlinux.h position
* update cargo build scripts to use the per-arch vmlinux.h for generating bindings
* keep the original ClangInfo refactoring changes

### CODE REVIEW

**Comment by minosfuture** [comment]
I made sure checking `meson build`, besides `cargo build`. The latter is also covered by CI though.
Please review again, @htejun @hodgesds @likewhatevs, thanks!

**Comment by sirlucjan** [comment]
Compiled without issue. Thanks.

**Comment by multics69** [comment]
I got the follow error. Is it ony me?

```
╰─ meson compile -C build -j32                                                                                                                                                                                                                                                 ─╯ 
INFO: autodetecting backend as ninja                                                                                                                                                                                                                                              
INFO: calculating backend command to run: /usr/bin/ninja -C /home/changwoo/Downloads/scx/build -j 32                                                                                                                                                                              
ninja: Entering directory `/home/changwoo/Downloads/scx/build'                                                                                                                                                                                                                    
[7/36] Generating 'scheds/c/scx_userland.p/scx_userland.bpf.o'                                                                                                                                                                                                                    
FAILED: scheds/c/scx_userland.p/scx_userland.bpf.o                                                                                                                                                                                                                                
/usr/bin/clang -g -O2 -Wall -Wno-compare-distinct-pointer-types -D__TARGET_ARCH_x86 -mcpu=v3 -mlittle-endian '-idirafter /usr/lib/clang/18/include' '-idirafter /usr/local/include' '-idirafter /usr/include' -target bpf -I /home/changwoo/Downloads/scx/build/libbpf/src/usr/include -I /home/changwoo/Downloads/scx/build/libbpf/include/uapi -I /home/changwoo/Downloads/scx/scheds/include -I /home/changwoo/Downloads/scx/scheds/include/vmlinux -I /home/changwoo/Downloads/scx/scheds/include/bpf-compat -c ../scheds/c/scx_userland.bpf.c -o scheds/c/scx_userland.p/scx_userland.bpf.o
In file included from ../scheds/c/scx_userland.bpf.c:23:
/home/changwoo/Downloads/scx/scheds/include/scx/common.bpf.h:14:10: fatal error: 'vmlinux.h' file not found
   14 | #include "vmlinux.h"
      |          ^~~~~~~~~~~

```

**Comment by minosfuture** [comment]
> I got the follow error. Is it ony me?
> 
> ```
> ╰─ meson compile -C build -j32                                                                                                                                                                                                                                                 ─╯ 
> INFO: autodetecting backend as ninja                                                                                                                                                                                                                                              
> INFO: calculating backend command to run: /usr/bin/ninja -C /home/changwoo/Downloads/scx/build -j 32                                                                                                                                                                              
> ninja: Entering directory `/home/changwoo/Downloads/scx/build'                                                                                                                                                                                                                    
> [7/36] Generating 'scheds/c/scx_userland.p/scx_userland.bpf.o'                                                                                                                                                                                                                    
> FAILED: scheds/c/scx_userland.p/scx_userland.bpf.o                                                                                                                                                                                                                                
> /usr/bin/clang -g -O2 -Wall -Wno-compare-distinct-pointer-types -D__TARGET_ARCH_x86 -mcpu=v3 -mlittle-endian '-idirafter /usr/lib/clang/18/include' '-idirafter /usr/local/include' '-idirafter /usr/include' -target bpf -I /home/changwoo/Downloads/scx/build/libbpf/src/usr/include -I /home/changwoo/Downloads/scx/build/libbpf/include/uapi -I /home/changwoo/Downloads/scx/scheds/include -I /home/changwoo/Downloads/scx/scheds/include/vmlinux -I /home/changwoo/Downloads/scx/scheds/include/bpf-compat -c ../scheds/c/scx_userland.bpf.c -o scheds/c/scx_userland.p/scx_userland.bpf.o
> In file included from ../scheds/c/scx_userland.bpf.c:23:
> /home/changwoo/Downloads/scx/scheds/include/scx/common.bpf.h:14:10: fatal error: 'vmlinux.h' file not found
>    14 | #include "vmlinux.h"
>       |          ^~~~~~~~~~~
> ```

Somehow yours is still using `-I /home/changwoo/Downloads/scx/scheds/include/vmlinux`, which should be `-I /home/changwoo/Downloads/scx/scheds/include/arch/{your arch, e.g. x86}` instead. Could you do a meson clean and see if this still persists? @multics69

File: rust/scx_utils/src/builder.rs [hunk 1]
```diff
@@ -35,6 +38,20 @@ impl Builder {
...
         let out_dir = env::var("OUT_DIR").unwrap();
+        let clang = ClangInfo::new().unwrap();
+        let kernel_target = clang.kernel_target().unwrap();
+        let vmlinux_h = Path::new(&BPF_H)
+            .join("arch")
+            .join(kernel_target)
+            .join("vmlinux.h")
+            .to_str()
+            .ok_or(anyhow!(
+                "{:?}/arch/vimlinux.h can't be converted to str",
```
Comment (htejun) [inline rust/scx_utils/src/builder.rs:49]: typo `vimlinux` -> `vmlinux`.

Comment (htejun) [reply rust/scx_utils/src/builder.rs:49]: Also, I think you can just `.unwrap()` here. This shouldn't happen and if it does, backtrace should have more than enough information for debugging.

**Comment by htejun** [comment]
Generally looks great to me.

**Comment by minosfuture** [comment]
> BTW, how are you generating `vmlinux.h` for all the different archs?

https://gist.github.com/minosfuture/e474dd32a93267e357b2ae6acbdf14c8#file-gen_vmlinux_h-sh

I used this script, with command `gen_vmlinux_h.sh /path/to/linux 5b7c893ed5ed0fc1cbe28c0e3296a6fb45698486`

File: rust/scx_utils/src/builder.rs (hunk 1 above)
Comment (minosfuture) [reply rust/scx_utils/src/builder.rs]: updated

**Comment by multics69** [comment]
@minosfuture I tested again with a fresh repo. It compiles well both in debug and release builds. Thank you for all the work!

BTW, I think file-gen_vmlinux_h.sh should be included in the `script` folder. What do you think?

**Comment by likewhatevs** [comment]
> BTW, I think file-gen_vmlinux_h.sh should be included in the script folder

I like this because one of the things I mean to do (sometime, probably around when 6.12 is released) is to:
(a) automate generating vmlinux.h's against the sched_ext tree's for-next branch (for development).
(b) automate generating vmlinux.h's against releases (i.e. 6.12).

Not sure how all that will work yet (maybe an extra CI job to test against the latest release and another CI job to generate commits add  vmlinux.h's), but this would be used heavily in all that.

**Comment by htejun** [comment]
> @minosfuture I tested again with a fresh repo. It compiles well both in debug and release builds. Thank you for all the work!
> 
> BTW, I think file-gen_vmlinux_h.sh should be included in the `script` folder. What do you think, @htejun ?

Yes, please.

**Comment by minosfuture** [comment]
#829 adds the script `gen_vmlinux_h.sh`

**Pull request merged**